### PR TITLE
Python 3

### PR DESCRIPTION
**Pull request by ALescoulie**
Replaced yield tests. Removed type annotations. Added more compatibility code for python 2.

### CODE REVIEW

**Comment by ALescoulie** [comment]
Just realized there was a small error in formatting a string in conifg, fixed it in a commit.

**Comment by ALescoulie** [comment]
With the fix to GromacsWrapper only 10 tests of 73 are failing, Solvate wet octanol, and the two analysis tests

**Comment by ALescoulie** [comment]
With the fix to GromacsWrapper only 10 tests of 73 are failing, solvation of wet octanol, and the two analysis tests

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        python -m pip install --no-deps .
+    - name: Run tests
+      shell: bash -l {0}
+      run: |
+        pytest -v --disable-pytest-warnings --durations=20 --low-performance --cov=gromacs --cov-report=xml ./tests
+    - name: Codecov
+      uses: codecov/codecov-action@v1
+      with:
+        file: ./coverage.xml
+        fail_ci_if_error: true
```
Comment (ALescoulie) [inline .github/workflows/CI.yml:87]: Started CI file similar to one by @orbeckst for gromacswrapper, work in progress.

File: mdpow/equil.py [hunk 2]
```diff
@@ -206,10 +211,23 @@ def __init__(self, molecule=None, **kwargs):
                                          box=self.solventmodel.coordinates,
                                          distance=distance)
 
-            self.filename = filename or self.solvent_type+'.simulation'
+            self.filename = filename or self.solvent_type + '.simulation'
 
         super(Simulation, self).__init__(**kwargs)
 
+    @staticmethod
+    def load_pickle():
```
Comment (ALescoulie) [inline mdpow/equil.py:219]: For compatibility between 2 and 3, could probably also use six

File: mdpow/tests/test_forcefields.py [hunk 3]
```diff
@@ -71,29 +72,32 @@ class TestWatermodels(object):
...
+
+        has_identifier(expected)
+        itp_in_top(expected)
+        coordinates_in_top(expected)
 
-        for identifier in self.watermodels:
-            yield has_identifier, identifier
-            yield itp_in_top, identifier
-            yield coordinates_in_top, identifier
+    def test_watermodelsdat(self):
```
Comment (ALescoulie) [inline mdpow/tests/test_forcefields.py:96]: Adding parameterized tests to replace yield tests

File: mdpow/tests/testing_resources/manifest.yml [hunk 4]
```diff
@@ -1,4 +1,4 @@
-path: mdpow/tests/testing_resources/states
+path: /nfs/homes3/alescoul/MDPOW_dev/mdpow/tests/testing_resources/states
```
Comment (ALescoulie) [inline mdpow/tests/testing_resources/manifest.yml:1]: Was having issues finding path, will need to replace.

File: mdpow/fep.py [hunk 5]
```diff
@@ -238,22 +253,33 @@ def load(cfg, section):
...
+
+        return FEPschedule((key, getter(keytype, section, key)) for key, keytype in keys.items()
                            if getter(keytype, section, key) is not None)
 
     def __deepcopy__(self, memo):
-        x = type(self)()
-        for k,v in self.iteritems():
-            x[k] = copy.deepcopy(v, memo)
-        return x
+        if sys.version_info.major == 3:
```
Comment (ALescoulie) [inline mdpow/fep.py:270]: iteritems() method not available in python 3, Iterable not variable in python 2, would like more elegant solution, but this should function.

File: mdpow/restart.py [hunk 6]
```diff
@@ -179,6 +186,17 @@ def __init__(self, *args, **kwargs):
             self.journal = Journal(self.protocols)
         super(Journalled, self).__init__(*args, **kwargs)
 
+    @staticmethod
+    def get_pickle():
+        if sys.version_info.major == 3:
+            import _pickle as pickle
```
Comment (ALescoulie) [inline mdpow/restart.py:192]: Ugly compatibility code probably could use six

File: mdpow/config.py [hunk 7]
```diff
@@ -92,35 +92,42 @@
...
 
 #: Locations of default run input files and configurations.
 defaults = {
     "runinput": resource_filename(__name__, "templates/runinput.yml"),
-    }
+}
+
 
-def merge_dicts(user, default):
+# TODO test merge_dicts
```
Comment (ALescoulie) [inline mdpow/config.py:116]: replaced with recursive method for merging dict containing multiple other dict, passed runinput

**Review by ALescoulie — COMMENTED** [review]
Current draft for update to python 3. Modified several outdated sections expecially around dictionary syntax. Replaced merge_dicts with recursive function with method not dependent on isiter.  Many changes just adding spacing to follow PEP8. Tests still need to be modified to run generally, at the moment some paths have been set for locations on workstation for sake of simplicity. Some Errors still occur in the analysis tests some of which appear to be associated with numkit integrate. 

Defiantly still a work in progress, hopefully nearing a functioning version.

File: mdpow/equil.py (hunk 2 above)
Comment (orbeckst) [reply mdpow/equil.py:219]: six is preferred — better use existing, tested libraries and six is standard and already used

File: mdpow/restart.py (hunk 6 above)
Comment (orbeckst) [reply mdpow/restart.py:192]: use six if possible — `from six.moves import pickle` probably, see docs and/or try it out interactively

File: mdpow/tests/test_Gsolv.py [hunk 8]
```diff
@@ -2,26 +2,25 @@
 import os
 import pybol
 import numpy as np
+import gromacs
 from gromacs.utilities import in_dir
+from mdpow.fep import *
+from mdpow.equil import *
```
Comment (orbeckst) [inline mdpow/tests/test_Gsolv.py:8]: avoid blanket imports — only import what you need or better, use the qualified names as before

File: mdpow/tests/test_analysis.py [hunk 9]
```diff
@@ -4,17 +4,18 @@
 
 import yaml
 import pybol
-
+import gromacs
 from numpy.testing import assert_array_almost_equal
 
 from six.moves import cPickle as pickle
 
-import mdpow.fep
```
Comment (orbeckst) [inline mdpow/tests/test_analysis.py:12]: leave as before with full paths as it makes it MUCH clearer where code comes from

File: mdpow/tests/test_analysis.py [hunk 10]
```diff
@@ -4,17 +4,18 @@
 
 import yaml
 import pybol
-
+import gromacs
 from numpy.testing import assert_array_almost_equal
 
 from six.moves import cPickle as pickle
```
Comment (orbeckst) [inline mdpow/tests/test_analysis.py:10]: example for pickle use

File: mdpow/tests/test_emin.py [hunk 11]
```diff
@@ -1,10 +1,13 @@
 import mdpow.equil
+import gromacs
 import tempdir as td
 import os
 
+
 class TestEnergyMinimization(object):
     
     def setup(self):
+        gromacs.config.set_gmxrc_environment('~/.conda/envs/MDPOW3/bin/GMXRC')
```
Comment (orbeckst) [inline mdpow/tests/test_emin.py:10]: should not be needed

File: mdpow/tests/test_fep.py [hunk 12]
```diff
@@ -1,27 +1,27 @@
 import numpy as np
 from numpy.testing import assert_array_almost_equal, assert_almost_equal
 from scipy import constants
-
-import mdpow
-import mdpow.fep
+import gromacs
+from mdpow.config import *
+from mdpow.fep import *
```
Comment (orbeckst) [inline mdpow/tests/test_fep.py:6]: keep full imports

File: mdpow/tests/test_fep.py [hunk 13]
```diff
@@ -50,7 +50,8 @@ class TestFEPschedule(object):
 
     def setup(self):
         # load default bundled configuration
-        self.cfg = mdpow.config.get_configuration()
+        gromacs.config.set_gmxrc_environment('~/.conda/envs/MDPOW3/bin/GMXRC')
```
Comment (orbeckst) [inline mdpow/tests/test_fep.py:53]: should not be needed

File: mdpow/tests/test_fep_script.py [hunk 14]
```diff
@@ -3,18 +3,19 @@
...
 from gromacs.utilities import in_dir
-
+import gromacs
 from mdpow.run import fep_simulation
 from mdpow.config import get_configuration
 
+
 class TestFEPScript(object):
     def setup(self):
+        gromacs.config.set_gmxrc_environment('~/.conda/envs/MDPOW3/bin/GMXRC')
```
Comment (orbeckst) [inline mdpow/tests/test_fep_script.py:13]: should not be needed

File: mdpow/tests/test_emin.py (hunk 11 above)
Comment (ALescoulie) [reply mdpow/tests/test_emin.py:10]: I'll try without it, I was having some issues with it not finding the conda GMXRC, mostly a temporary solution to get tests running so I could evaluate my code. I defiantly need to remove some of the modifications I made while trying to troubleshoot import errors.

File: mdpow/tests/test_analysis.py (hunk 9 above)
Comment (ALescoulie) [reply mdpow/tests/test_analysis.py:12]: I'll make that replacement in fep and equil in a new commit.

**Comment by ALescoulie** [comment]
@orbeckst I looked at your comments and to address the concerns will add six.moves to the files currently using cPickle or _pickle. I'll also go back and try to reverse some of the changes I made while trying to troubleshoot getting tests to run for me. They were a bit finicky to get working in particular surrounding sourcing GMXRC and importing the modules.

**Comment by ALescoulie** [comment]
Just pushed 3 commits addressing @orbeckst comments on the code. Fixed the test imports and switched to six.moves, also removed the set_gmxrc_environment worked without it.

**Comment by orbeckst** [comment]
As discussed offline: start from a recent develop (at least 355dbe8ecb5ccd2d4a3390a65c212f470a899df9) and only add code that is directly necessary for Py 2/3 compatibility. I took your changes for the tests (yield replacement and added them to PR #164 e.g. in 86f563829c440929a51e90f051e30cd06ef88509 ).

**Comment by orbeckst** [comment]
Close this PR because it is superseded by PR #166.

**Pull request closed without merging**